src/coreclr/jit/optimizemaskconversions.cpp

```diff
@@ -0,0 +1,508 @@
+// must first convert from vector before using it. In many cases this creates unnecessary
+// conversions. For variables that live outside the scope of the current method then the
+// conversions are required to ensure correctness. However, for local variables where the
+// scope is local to the current method, then it is possible to keep the value as a mask,
+// by updating all definitions and uses.
+//
+// In the common case it is expected that uses of masks are consistent - once a variable is
+// created as a mask it will continue to be used and updated as a mask.
+//
+// In the uncommon case, a variable may be created in one type, used as another and/or
+// updated to a different type.
+//
+// For example (the conversion is implicit)
+//   vector<int> x = _ConvertMaskToVector_(CreateMask());
+//   x = Add(x, y);
+//
+// To optimize this, the pass searches every local variable definition (GT_STORE_LCL_VAR)
+// and use (GT_LCL_VAR). A weighting is calculated and kept in a hash table - one entry
+// for each lclvar number. The weighting contains two values. The first value is the count of
+// of every convert node for the var, each instance multiplied by the number of instructions
+// in the convert and the weighting of the block it exists in. The second value assumes the
+// local var has been switched to store as a mask and performs the same count. The switch
+// will count removes every existing convert and add a convert where there isn't currently
+// a convert.
+//
+// Once every definition and use has been parsed, the parsing runs again. At each step,
+// if the weighting for switching that var is lower than the current weighting then switch
+// to store as mask and add/remove conversions as required.
+//
+// Limitations:
+//
+// Local variables that are defined then immediately used just once may not be saved to a
+// store. Here a convert to to vector will be used by a convert to mask. These instances will
+// be caught in the lowering phase.
+//
+// This weighting does not account for re-definition. A variable may first be created as a
+// mask used as such, then later in the method defined as a vector and used as such from
+// then on. This can be worked around at the user level by encouraging users not to reuse
+// variable names.
+//
+// Returns:
+//    Suitable phase status
+//
+PhaseStatus Compiler::fgOptimizeMaskConversions()
+{
+#if defined(TARGET_ARM64)
+
+    if (opts.OptimizationDisabled())
+    {
+        JITDUMP("Skipping. Optimizations Disabled\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+#if defined(DEBUG)
+    if (JitConfig.JitDoOptimizeMaskConversions() == 0)
+    {
+        JITDUMP("Skipping. Disable by config option\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+#endif
+
+    if (!compMaskConvertUsed)
+    {
+        JITDUMP("Skipping. There are no converts of locals\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    MaskConversionsWeightTable weightsTable(getAllocator());
```

Comment:
Can you introduce a new memory kind in `compmemkind.h` and use it for this allocator? `getAllocator` defaults to `CMK_DebugOnly` which we should avoid using.
```suggestion
    MaskConversionsWeightTable weightsTable(getAllocator(CMK_MaskConversionOpt));
```